client/views/view_units.cpp

```diff
@@ -0,0 +1,466 @@
+  total_food = 0;
+  max_row = 0;
+
+  get_units_view_data(unit_entries, &entries_used);
+  for (i = 0; i < entries_used; i++) {
+    struct unit_view_entry *pentry = unit_entries + i;
+    struct unit_type *putype = pentry->type;
+    cid id;
+
+    auto sprite = get_unittype_sprite(tileset, putype, direction8_invalid());
+    id = cid_encode_unit(putype);
+
+    ui.units_widget->insertRow(max_row);
+    for (j = 0; j < 7; j++) {
+      item = new QTableWidgetItem;
+      switch (j) {
+      case 0:
+        // Unit type image sprite
+        if (sprite != nullptr) {
+          item->setData(Qt::DecorationRole, sprite->scaledToHeight(h));
+        }
+        item->setData(Qt::UserRole, id);
+        item->setTextAlignment(Qt::AlignLeft);
+        item->setText(utype_name_translation(putype));
+        break;
+      case 1:
+        // Upgradable
+        item->setData(Qt::DisplayRole, 0);
+        break;
+      case 2:
+        // In Progress
+        item->setData(Qt::DisplayRole, 0);
+        break;
+      case 3:
+        // Active
+        item->setData(Qt::DisplayRole, pentry->count);
+        total_count += pentry->count;
+        break;
+      case 4:
+        // Shield upkeep
+        if (pentry->shield_cost == 0) {
+          item->setText("-");
+          item->setTextAlignment(Qt::AlignCenter);
+        } else {
+          item->setData(Qt::DisplayRole, pentry->shield_cost);
+        }
+        total_shield += pentry->shield_cost;
+        break;
+      case 5:
+        // Food upkeep
+        if (pentry->food_cost == 0) {
+          item->setText("-");
+          item->setTextAlignment(Qt::AlignCenter);
+        } else {
+          item->setData(Qt::DisplayRole, pentry->food_cost);
+        }
+        total_food += pentry->food_cost;
+        break;
+      case 6:
+        // Gold upkeep
+        if (pentry->gold_cost == 0) {
+          item->setText("-");
+          item->setTextAlignment(Qt::AlignCenter);
+        } else {
+          item->setData(Qt::DisplayRole, pentry->gold_cost);
+        }
+        total_gold += pentry->gold_cost;
+        break;
+      }
+      item->setTextAlignment(Qt::AlignVCenter | Qt::AlignHCenter);
+      ui.units_widget->setItem(max_row, j, item);
+    }
+    max_row++;
+  }
+
+  // max_row ++;
+  ui.units_widget->setRowCount(max_row);
+  ui.units_widget->insertRow(max_row);
+  for (j = 0; j < 7; j++) {
+    item_totals = new QTableWidgetItem;
+    switch (j) {
+    case 0:
+      // Unit type
+      item_totals->setTextAlignment(Qt::AlignRight);
+      item_totals->setText(_("---------------\nTotals:"));
+      break;
+    case 1:
+      // Upgradable
+      item_totals->setTextAlignment(Qt::AlignCenter);
+      item_totals->setText(_("-------\n0"));
+      // item->setData(Qt::DisplayRole, 0);
+      break;
+    case 2:
+      // In Progress
+      item_totals->setTextAlignment(Qt::AlignCenter);
+      item_totals->setText(_("-------\n0"));
+      // item->setData(Qt::DisplayRole, 0);
+      break;
```

Comment:
Here a loop would make sense since the exact same code is repeated many times. But would require putting the totals in an array...